### DISCUSSION TITLE
Local or internalised library

### DISCUSSION THREAD

**Discussion by toniosterlund** [visose/Robots#147] [Q&A] [answered]
Hi,
I'm testing the plugin and regarding the Robot Library, as far as I understand, in order to add your own robot, you need to either
1) Download an existing library and modify it locally
2) Upload a new library to github, where it is publicly available
Did I understand correctly?

However, working in a commercial setting, I looking for a possibility to define the robot completely locally, and preferably so, that the robot is internalised into the gh-file itself (much like in Kuka|prc). Is there a possibility to have a local library (with unique names), maybe stored to a server location, other than User/Robots/ ?

**Comment by visose** [comment, marked as answer]
Hi @toniosterlund, as you say, if you don't want your robot library to be publicly available, currently the solution would be to create locally the 3dm/xml files placed under the User/Robots/ folder (Documents/Robots in Windows) and distribute these files to other machines, asking the users to copy them to this same location. If you change these files, they will have to copy them again.

We could add additional folder locations to check for libraries, these could point to a network drive. We could also add a feature to load a robot system from internalized text and a list of meshes per mechanism.

If you are in a commercial setting, are finding the plugin useful, and can afford it, consider sponsoring the project to get your feature requests prioritized.

**Reply by toniosterlund** [reply]
Hi @visose , thank you for a speedy answer. For now, I can live with the library in the user document folder, though thematically it's a bit iffy to hack someone else's library to get your own setup.
I'm currently using Kuka|prc with about 2,500e/year, so there is a large motivation to sponsor you for your work. All aside Kuka|prc still works like a charm and has been very simple to set up - so all kudos to Johannes and co. But I have to say that from my brief testing, your Robots plugin seems to work very well as well - I just need to keep testing.

Semi-related question: Where do I define Tool and Base numbers?

**Reply by visose** [reply]
> though thematically it's a bit iffy to hack someone else's library to get your own setup

I'm not sure exactly what you mean. You can create your own xml and 3dm files from scratch, but to understand how to format the xml file, you can check the ones in this repo as an example. An example is sometimes easier to understand than a description (but ideally you want both).

> Where do I define Tool and Base numbers?

Each robot manufacturer use slightly different terminology to refer to different coordinate systems, since you are using KUKA:
<img width="341" alt="image" src="https://user-images.githubusercontent.com/5129614/157501220-ceebbba2-cc5a-4090-af6c-f77e2119112c.png">

"Base CS" in the diagram above is equivalent to the "Frame" parameter in the plugin. Each target will have a Frame associated with it.

When creating the xml file, each mechanism has a <Base ..> element that is equivalent to "Robot CS" in the diagram above. It is defined as 7 numbers (xyz + 4 quaternion values). This should be set to match the value configured in the controller, which by default is the origin, but could have been set to something else. The plugin has a component to convert kuka's XYZ+ABC values to xyz + quaternions.

There is also an input in the create robot system component to define a different robot base (Robot CS), without having to change the configured value in the controller. Use this if you want the origin of the Rhino document to not be the base of the robot. This works by transforming the coordinates of the Frame parameters (Base CS) in the generated KRL code.

For the tool, you can use the create tool component to define one or more tools in the Grasshopper document, then attach them to each target (like Frame). You can also use the load tool component if you want to define the tool in xml rather than the grasshopper file. For an example of this check the Taiwan Tech library.

**Reply by toniosterlund** [reply]
Hi @visose 
I mean, we have the Tools and Bases defined in the robot itself, and are only referred through numbers. I can setup their locations in Gh to match their real-world counterparts, but I would not like to pass their frames explicitly in scr.

This is an example of how the file header is defined currently.
![image](https://user-images.githubusercontent.com/101263308/157514476-b4a5224b-8d29-4dbd-8ac8-6cb5b280d698.png)

**Reply by visose** [reply]
> Hi @visose I mean, we have the Tools and Bases defined in the robot itself, and are only referred through numbers.

It was decided early on to avoid relying on user data stored in the robot controller for multiple reasons. The same robots were used by multiple teams and they might trample over each other when modifying data in the controller. Due to the nature of how robots are used, each team usually provided their own tool and base, so these were considered user data. 
Having the Grasshopper file be the "single source of truth" removes ambiguity and possible mistakes where the simulation might not match what the robot actually does.

> but I would not like to pass their frames explicitly in scr

Why not?

**Reply by toniosterlund** [reply]
We have the situation vice versa. A couple of people produce toolpaths and the robot has its own operator(s). Few hundred kilometers of physical distance, so I am not anywhere near when the robot is being operated. This means that the operator makes small adjustments to the base all the time in order to match X or Z to the base material being milled. We try to keep the base matching on the simulation end, but a few millimeter mismatches do not matter - as it's just simulation. And in the end, it's the tool's CP that matters.

I was looking through the Code output of the 'Create Program' component, and I suppose I could extract and replace the header from the  {2} branch and just save that as an src..? Just need to construct my own type of header and save the merged strings as a file. Both would be completely doable.
![image](https://user-images.githubusercontent.com/101263308/157647626-239535f0-79e9-4e8a-b94d-9ee5e33684e7.png)

But I actually just realised that the first PTP is given as axis values, when the subsequent ones are XYZabc. Why is that? Is the first automatically considered as Home position?

**Reply by toniosterlund** [reply]
Seems I need to study the plugin further, as the 'Simple example' file gives a neater code on LIN.
![image](https://user-images.githubusercontent.com/101263308/157650003-4829b491-54d4-43ab-88ab-8a831135d521.png)

**Reply by visose** [reply]
> I suppose I could extract and replace the header from the {2} branch and just save that as an src

What you can do is:
- Modify the code output from the "create program" component. Use text replacement grasshopper components to replace the lines starting with $TOOL, $LOAD, $BASE, etc. (don't separate into two, just replace the lines you need and should remain a working program)
- Feed the modified code to a  "create custom program" component.
- Use the output of the custom program with the save file component.

> just realised that the first PTP is given as axis values, when the subsequent ones are XYZabc. Why is that? 

Yes, the Warnings output should mention that it was changed to a joint target. The reason for this is to make sure the program starts from an unambiguous pose, irrespective or where the robot was before starting the program. With other types of target, the robot might reach it in different ways depending on its previous position.

**Reply by visose** [reply]
>  'Simple example' file gives a neater code on LIN.

The reason for so many speed changes in the PTP motions is because the plugin tries to approximate the speed set in the Grasshopper targets, which is defined in mm/s, on joint movements as well, this means setting a different leading axis speed for each target.

**Reply by toniosterlund** [reply]
Ok, thank you for your answers!

**Comment by visose** [comment]
@toniosterlund, loading tools and frames directly from the controller has been added to the latest version. See this comment for details: https://github.com/visose/Robots/discussions/154#discussioncomment-2851005

**Comment by IBR4NX** [comment]
هلا كيف يمكنني أن اتعلم ؟
